src/sc_client/session.py

```diff
@@ -66,91 +84,110 @@ def _emit_callback(event_id: int, elems: list[int]) -> None:
-        set_reconnection(url)
+def set_connection(url: str) -> None:
+    establish_connection(url)
+
+
+def is_connected() -> bool:
+    check_request = json.dumps({"type": "healthcheck"})
+    _ScClientSession.ws_app.send(check_request)
+
+    while not _ScClientSession.last_healthcheck_answer and _ScClientSession.is_open:
+        time.sleep(SERVER_ANSWER_CHECK_TIME)
+
+    result = _ScClientSession.last_healthcheck_answer == "OK"
+    _ScClientSession.last_healthcheck_answer = None
+    return result
 
 
 def establish_connection(url) -> None:
     def run_in_thread():
-        _ScClientSession.ws_app = websocket.WebSocketApp(
-            url,
-            on_message=_on_message,
-            on_error=_on_error,
-            on_close=_on_close,
-        )
-        _ScClientSession.ws_app.run_forever()
+        if not _ScClientSession.ws_app or _ScClientSession.ws_app.url != url:
+            _ScClientSession.ws_app = websocket.WebSocketApp(
+                url,
+                on_message=_on_message,
+                on_error=_on_error,
+                on_close=_on_close,
+            )
+            logger.info(f"Sc-server socket: {_ScClientSession.ws_app.url}")
+        if not _ScClientSession.is_open:
+            try:
+                _on_open(_ScClientSession.ws_app)
+                logger.info("Set new connection")
+                _ScClientSession.ws_app.run_forever()
+            except websocket.WebSocketException as e:
+                _on_error(_ScClientSession.ws_app, e)
 
     client_thread = threading.Thread(target=run_in_thread, name="sc-client-session-thread")
     client_thread.start()
     time.sleep(SERVER_ESTABLISH_CONNECTION_TIME)
-    logger.debug("Trying to establish connection: %s", "successful" if is_connection_established() else "unsuccessful")
-
-
-def set_reconnection(url: str) -> None:
-    def run_in_thread():
-        while _ScClientSession.do_reconnect:
-            time.sleep(SERVER_ESTABLISH_CONNECTION_TIME)
-            if not is_connection_established():
-                logger.debug("No connection, trying to reconnect")
-                establish_connection(url)
-                if not is_connection_established():
-                    logger.debug("Still no connection, waiting")
-                    time.sleep(_ScClientSession.reconnecion_time)
-
-    reconnection_thread = threading.Thread(target=run_in_thread, name="reconnection-thread")
-    reconnection_thread.start()
-
-
-def disable_reconnection() -> None:
-    _ScClientSession.do_reconnect = False
 
 
 def close_connection() -> None:
     try:
         _ScClientSession.ws_app.close()
-    except AttributeError:
-        pass
-    _ScClientSession.ws_app = None
+    except AttributeError as e:
+        _on_error(_ScClientSession.ws_app, e)
     logger.debug("Disconnected")
 
 
 def receive_message(command_id: int) -> Response:
     response = None
-    while not response:
+    while not response and _ScClientSession.is_open:
```

Comment:
We're returning `None` right away if the connection isn't open. Maybe we should raise an Exception in this case?

Reply:
If there is no connection when client tries to get a response from his server, then this means that the server will never return again. Therefore, we must restore the connection, if it appeared, send a request to the server and receive a response. If it fails to get a response the second time, it will return None. Perhaps in this case you need to throw the  sugested error, but not in this place